### PR TITLE
Add distillation for segmentation

### PR DESCRIPTION
**Pull request by juncaipeng**
* Add segmentation distillation api, demo and docs
* Update the Kl loss and cross entropy loss
* Refine the padding in the Resnet_vd

### CODE REVIEW

File: configs/quick_start/deeplabv3p_resnet50_os8_optic_disc_512x512_1k.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
+batch_size: 4
```
Comment (michaelowenliu) [inline configs/quick_start/deeplabv3p_resnet50_os8_optic_disc_512x512_1k.yml]: 不放在quick start里面，放在deeplabv3p里面

File: docs/slim/distill/distill.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
+# 模型蒸馏教程
+
+## 1 简介
+
+模型蒸馏基于”教师-学生模型“的思想，使用大的教师模型指导小的学生模型进行训练，是一种常见的模型压缩方法。相比于单独训练学生模型，模型蒸馏通常可以提升学生模型的精度。
+
+常规的模型训练中，模型前向计算的输出和真实label计算得到常规loss，再进行梯度反向传播。
+常见的模型蒸馏训练中，教师模型只有前向计算，学生模型有前向计算和反向传播，有多个loss指导学生模型进行训练：学生模型前向计算的输出和真实label计算得到常规loss；学生模型前向计算的输出和教师模型前向计算的输出计算得到蒸馏loss。
+更多模型蒸馏的介绍，请参考[Survey](https://arxiv.org/abs/2006.05525)和[博客](https://cloud.tencent.com/developer/article/1763873)。
```
Comment (michaelowenliu) [inline docs/slim/distill/distill.md]: 这个博客删除吧，不是我们官方的

File: docs/slim/distill/distill.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+修改教师模型的config文件（`PaddleSeg/configs/quick_start/deeplabv3p_resnet50_os8_optic_disc_512x512_1k.yml`），将文件中最后一行pretrained字段设置为”训练教师模型”步骤中的权重路径，如下所示。
+```
+model:
+  type: DeepLabV3P
+  backbone:
+    type: ResNet50_vd
+    output_stride: 8
+    multi_grid: [1, 2, 4]
+    pretrained: https://bj.bcebos.com/paddleseg/dygraph/resnet50_vd_ssld_v2.tar.gz
```
Comment (michaelowenliu) [inline docs/slim/distill/distill.md]: 这里可以把backbone的pretrained删掉，以免用户疑惑

File: docs/slim/distill/distill.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+  num_classes: 2
+  backbone_indices: [0, 3]
+  aspp_ratios: [1, 12, 24, 36]
+  aspp_out_channels: 256
+  align_corners: False
+  pretrained: output/deeplabv3p_resnet50/best_model/model.pdparams
+```
+
+学生模型DeepLabV3P_ResNet18_vd的config文件在`PaddleSeg/configs/quick_start/deeplabv3p_resnet18_os8_optic_disc_512x512_1k.yml`，其中除了常规loss，还新增了distill_loss，如下所示。常规loss是配置学生模型输出和真实label的损失计算，distill_loss是配置学生模型输出和教师模型输出的损失计算，type表示loss类型，coef是loss的比例系数。
```
Comment (michaelowenliu) [inline docs/slim/distill/distill.md]: 可以在config文件里面的名字加个teacher和student字样，这样更清楚了

File: docs/slim/distill/distill.md [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+       --use_vdl \
+       --save_interval 250 \
+       --num_workers 3 \
+       --seed 0 \
+       --save_dir output/deeplabv3p_resnet18_distill
+```
+
+在蒸馏训练中，使用教师模型配置文件中model配置信息创建教师模型，使用学生模型配置文件中model配置信息创建学生模型，使用学生模型中dataset、loss、optimizer等配置信息执行训练。
+
+注意，蒸馏训练会加载两个模型，显存占用较大，所以大家需要根据实际情况调整batch_size。
```
Comment (michaelowenliu) [inline docs/slim/distill/distill.md]: 这里给出一个显存的参考，比如batch_size=2，对应的显存多少

File: paddleseg/cvlibs/config.py [hunk 6]
```diff
@@ -238,29 +257,27 @@ def loss(self) -> dict:
             raise ValueError(
                 'Loss config should contain keys of "types" and "coef"')
 
-        if not self._losses:
```
Comment (michaelowenliu) [inline paddleseg/cvlibs/config.py:241]: self._losses被删掉了吗？

File: paddleseg/models/losses/kl_loss.py [hunk 7]
```diff
@@ -21,17 +22,7 @@
...
-    Code referenced from:
-
-    https://github.com/peterliht/knowledge-distillation-pytorch/blob/master/model/net.py
-
-
-
-    Compute the knowledge-distillation (KD) loss given outputs, labels.
-
-    "Hyperparameters": temperature and alpha
+    The implementation of Kullback-Leibler divergence Loss.
```
Comment (michaelowenliu) [inline paddleseg/models/losses/kl_loss.py:25]: 这里需要给出一个KL loss的简单解释，或者reference paper

File: paddleseg/models/losses/kl_loss.py [hunk 8]
```diff
@@ -40,12 +31,49 @@ class KLLoss(nn.Layer):
...
+        loss = self.kl_loss(logit_1, logit_2)
+        loss = loss * self.temperature * self.temperature
+
+        if label is None:
+            avg_loss = paddle.mean(loss)
+        else:
+            mask = label != self.ignore_index
+            mask = paddle.cast(mask, 'float32')
+            mask = paddle.unsqueeze(mask, axis=1)
+            label.stop_gradient = True
```
Comment (michaelowenliu) [inline paddleseg/models/losses/kl_loss.py:75]: 这个label的作用只是mask的话，参数名可以改成mask吗

File: slim/distill/distill_config.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+
+def prepare_distill_adaptor():
+    """
+    Prepare the distill adaptors for student and teacher model.
+    The adaptors set the intermediate feature tensors that used for distillation.
+    """
+
+    class StudentAdaptor(AdaptorBase):
+        def mapping_layers(self):
```
Comment (michaelowenliu) [inline slim/distill/distill_config.py:25]: 这是个空函数？

File: slim/distill/distill_config.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+            mapping_layers = {}
+            # mapping_layers['hidden_0'] = 'layer_name'
+            if self.add_tensor:
+                # mapping_layers["hidden_0"] = self.model.logit_list
+                pass
+            return mapping_layers
+
+    class TeatherAdaptor(AdaptorBase):
+        def mapping_layers(self):
```
Comment (michaelowenliu) [inline slim/distill/distill_config.py:34]: 这也是空函数？

File: slim/distill/distill_config.py [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    '''
+    distill_config = [{
+        's_feature_idx': 0,
+        't_feature_idx': 0,
+        'feature_type': 'hidden',
+        'loss_function': 'SegChannelwiseLoss',
+        'weight': 1.0
+    }]
+    '''
+    distill_config = {}
```
Comment (michaelowenliu) [inline slim/distill/distill_config.py:58]: 最后返回空吗？

**Review by michaelowenliu — COMMENTED** [review]
add a few comments.

File: configs/quick_start/deeplabv3p_resnet50_os8_optic_disc_512x512_1k.yml (hunk 1 above)
Comment (juncaipeng) [reply configs/quick_start/deeplabv3p_resnet50_os8_optic_disc_512x512_1k.yml]: 放到quick_start中，加入teacher和student名字

File: docs/slim/distill/distill.md (hunk 2 above)
Comment (juncaipeng) [reply docs/slim/distill/distill.md]: Done

File: docs/slim/distill/distill.md (hunk 3 above)
Comment (juncaipeng) [reply docs/slim/distill/distill.md]: Done

File: docs/slim/distill/distill.md (hunk 4 above)
Comment (juncaipeng) [reply docs/slim/distill/distill.md]: Done

File: docs/slim/distill/distill.md (hunk 5 above)
Comment (juncaipeng) [reply docs/slim/distill/distill.md]: Done

File: paddleseg/cvlibs/config.py (hunk 6 above)
Comment (juncaipeng) [reply paddleseg/cvlibs/config.py:241]: 没有啊，在这里：https://github.com/PaddlePaddle/PaddleSeg/pull/1253/files#diff-2b12d13674290553151130755a31491793ecb5fb62b9552387deea3d22bace28R226

File: paddleseg/models/losses/kl_loss.py (hunk 7 above)
Comment (juncaipeng) [reply paddleseg/models/losses/kl_loss.py:25]: 放了一个Wikipedia链接，这是比较常见的概念。

File: paddleseg/models/losses/kl_loss.py (hunk 8 above)
Comment (juncaipeng) [reply paddleseg/models/losses/kl_loss.py:75]: 这个输入是label，基于label和ignore_index才计算mask，所以还是用label吧。

File: slim/distill/distill_config.py (hunk 9 above)
Comment (juncaipeng) [reply slim/distill/distill_config.py:25]: 嗯，这个中间结果蒸馏的方法，后续再补充，注释的这些是一个示例。

File: slim/distill/distill_config.py (hunk 10 above)
Comment (juncaipeng) [reply slim/distill/distill_config.py:34]: 如上

File: slim/distill/distill_config.py (hunk 11 above)
Comment (juncaipeng) [reply slim/distill/distill_config.py:58]: 如上

**Review by michaelowenliu — APPROVED** [review]
LGTM
